tests/foreman/api/test_permission.py

```diff
@@ -5,34 +5,56 @@
 tested can be found here: http://theforeman.org/api/apidoc/v2/permissions.html
 
 """
+import json
+import random
 import re
-from ddt import data as ddt_data, ddt
+
+from ddt import ddt
 from fauxfactory import gen_alphanumeric
 from itertools import chain
 from nailgun import client, entities
 from requests.exceptions import HTTPError
 from robottelo.common.constants import PERMISSIONS
 from robottelo.common.decorators import bz_bug_is_open, data, run_only_on
-from robottelo.common.helpers import get_nailgun_config, get_server_credentials
+from robottelo.common.helpers import (
+    get_nailgun_config,
+    get_server_credentials,
+    get_server_software,
+)
 from robottelo.test import APITestCase
 # (too-many-public-methods) pylint:disable=R0904
 
 
-#: e.g. ['Architecture', 'Audit', 'AuthSourceLdap', …]
-PERMISSION_RESOURCE_TYPES = [key for key in PERMISSIONS.keys()]
-#: e.g. ['view_architectures', 'create_architectures', 'edit_architectures', …]
-PERMISSION_NAMES = [
-    value for value in chain.from_iterable(PERMISSIONS.values())
-]
-
-
 @ddt
 class PermissionsTestCase(APITestCase):
     """Tests for the ``permissions`` path."""
+    @classmethod
+    def setUpClass(cls):
+        super(PermissionsTestCase, cls).setUpClass()
+        cls.permissions = PERMISSIONS.copy()
+        if get_server_software() == 'upstream':
+            cls.permissions[None] += cls.permissions.pop('DiscoveryRule')
```

Comment:
It will have just two lines. This is needed because the DiscoveryRule permissions on upstream are not associated with that resource type but with the null one. List or tuple this construction will remain the same.

A list will be better in the next step which some permissions need to be removed.